dsc/config.py

```diff
@@ -1,33 +1,61 @@
 import logging
 import os
+from collections.abc import Iterable
+from io import StringIO
 
 import sentry_sdk
 
 
-def configure_logger(logger: logging.Logger, *, verbose: bool) -> str:
-    if verbose:
-        logging.basicConfig(
-            format="%(asctime)s %(levelname)s %(name)s.%(funcName)s() line %(lineno)d: "
-            "%(message)s"
-        )
-        logger.setLevel(logging.DEBUG)
-        for handler in logging.root.handlers:
-            handler.addFilter(logging.Filter("dsc"))
-    else:
-        logging.basicConfig(
-            format="%(asctime)s %(levelname)s %(name)s.%(funcName)s(): %(message)s"
+class Config:
+    REQUIRED_ENV_VARS: Iterable[str] = [
+        "WORKSPACE",
+        "SENTRY_DSN",
+        "AWS_REGION_NAME",
+    ]
+
+    OPTIONAL_ENV_VARS: Iterable[str] = ["LOG_LEVEL"]
+
+    def __getattr__(self, name: str) -> str | None:
+        """Provide dot notation access to configurations and env vars on this class."""
+        if name in self.REQUIRED_ENV_VARS or name in self.OPTIONAL_ENV_VARS:
+            return os.getenv(name)
+        message = f"'{name}' not a valid configuration variable"
+        raise AttributeError(message)
+
+    def check_required_env_vars(self) -> None:
+        """Method to raise exception if required env vars not set."""
+        missing_vars = [var for var in self.REQUIRED_ENV_VARS if not os.getenv(var)]
+        if missing_vars:
+            message = f"Missing required environment variables: {', '.join(missing_vars)}"
+            raise OSError(message)
+
+    def configure_logger(
```

Comment:
This is a combination of `wiley-deposits` and our template config code, there's some repetition and I have a gut feeling there's a better way to do this but land on anything specific

Reply:
Here's a possible refactor, removing some redundancies:

```python
def configure_logger(
    self, logger: logging.Logger, stream: StringIO, *, verbose: bool
) -> str:
    logging_format_base = "%(asctime)s %(levelname)s %(name)s.%(funcName)s()"
    logger.addHandler(logging.StreamHandler(stream))

    if verbose:
        log_method, log_level = logger.debug, logging.DEBUG
        template = logging_format_base + " line %(lineno)d: %(message)s"
        for handler in logging.root.handlers:
            handler.addFilter(logging.Filter("dsc"))
    else:
        log_method, log_level = logger.info, logging.INFO
        template = logging_format_base + ": %(message)s"

    logger.setLevel(log_level)
    logging.basicConfig(format=template)
    logger.addHandler(logging.StreamHandler(stream))
    log_method(f"{logging.getLevelName(logger.getEffectiveLevel())}")

    return (
        f"Logger '{logger.name}' configured with level="
        f"{logging.getLevelName(logger.getEffectiveLevel())}"
    )
```

At a glance, what makes this hard to refactor are the tests:
- `test_configure_logger_not_verbose`
- `test_configure_logger_verbose`

which attempt to read from the stream handler that is setup.   That's what the `log_method, log_level = ` do above, is get the appropriate logging method to use for `log_method(f"{logging.getLevelName(logger.getEffectiveLevel())}")` which satisfies these tests.

My feeling: good enough for now, but globally I think we could improve our logging setups.  The presence of `logging.basicConfig()` is kind of a red flag unto itself.  But, I think this is mostly consistent with _other_ projects, so value in that.

Reply:
Great refactor, thanks! I was hoping this would prompt a larger conversation about the template/config class because I think it needs work as well, maybe a future DataEng meeting topic?